### PR TITLE
[improve][proxy] Support zero-copy of NIC to NIC on Proxy

### PR DESCRIPTION
**Pull request by coderzc**
Master Issue: #15631 

### Motivation

Currently, when the Proxy server transfer data will have four triggered the user mode and kernel mode of context switching, respectively is the read() / write() call and return when cut, it is inefficient for the Proxy.

There is an efficient way of reducing the times of copies of data and reduce CPU load. for linux system, we can use [`splice`](https://man7.org/linux/man-pages/man2/splice.2.html) system call.

### Modifications

* add a new config `proxyZeroCopyModeEnabled` in the proxy server, If enabled, we will Use [splice](https://netty.io/news/2015/05/07/4-0-28-Final.html#:~:text=This%20release%20includes%20support%20for%20splice%20when%20using%20the%20native%20epoll%20transport.%20Splicing%20allows%20to%20directly%20transfer%20bytes%20from%20one%20filedescriptor%20to%20another%20without%20copy%20data%20from%20kernel%20to%20user%2Dspace.%20This%20makes%20splice%20a%20perfect%20fit%20i.e%20when%20you%20want%20to%20write%20a%20TCP%20proxy.) system call instead of read() / write() for linux system.
* It should only work as `proxyLogLevel == 0`
* It can't work when the connection opens TLS 

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is already covered by existing tests

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `no-need-doc` 
(Please explain why)
  
- [x] `doc` 
(Your PR contains doc changes)

- [x] `doc-added`
(Docs have been already added)

### CODE REVIEW

File: conf/proxy.conf [hunk 1]
```diff
@@ -60,6 +60,9 @@ advertisedAddress=
 # Enable or disable the HAProxy protocol.
 haProxyProtocolEnabled=false
 
+# Enable or disable the zero-copy of NIC to NIC.
+zeroCopyModeEnabled=true
```
Comment (cbornet) [inline conf/proxy.conf:64]: zero-copy can mean a lot of things. Can we be more explicit here ?

Comment (coderzc) [reply conf/proxy.conf:64]: ok, I will add more descriptions.

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java [hunk 2]
```diff
@@ -109,6 +112,10 @@ public void connect(String brokerHostAndPort, InetSocketAddress targetBrokerAddr
         b.group(inboundChannel.eventLoop())
                 .channel(inboundChannel.getClass());
 
+        if (service.zeroCopyModeEnabled) {
+            b.option(EpollChannelOption.EPOLL_MODE, EpollMode.LEVEL_TRIGGERED);
```
[the remark this answers is not in the document]
Comment (coderzc) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:116]: ok, I will check `inboundChannel.getClass()` make sure it is `EpollSocketChannel` type

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java [hunk 3]
```diff
@@ -346,6 +353,14 @@ public void channelRead(final ChannelHandlerContext ctx, Object msg) throws Exce
                 }
                 inboundChannel.writeAndFlush(msg)
                         .addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
+
+                if (service.zeroCopyModeEnabled && service.proxyLogLevel == 0) {
+                    if (!ProxyConnection.isTlsChannel(ctx.channel()) && !ProxyConnection.isTlsChannel(inboundChannel)) {
```
[the remark this answers is not in the document]
Comment (coderzc) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java:358]: > what about caching this condition ? the "channel" is constant. we can save some cycles

Caching the `!ProxyConnection.isTlsChannel(inboundChannel)` ?

File: conf/proxy.conf (hunk 1 above)
Comment (coderzc) [reply conf/proxy.conf:65]: > I would name this like 'proxyZeroCopyModeEnabled'

ok, the name is more clear

File: pulsar-proxy/src/test/resources/proxy.conf [hunk 4]
```diff
@@ -60,6 +60,10 @@ advertisedAddress=
 # Enable or disable the HAProxy protocol.
 haProxyProtocolEnabled=false
 
+# Enable or disable the zero-copy of NIC to NIC. If enabled, we will use splice system call instead of read/write to
+# transport data of NIC to NIC, it only work when `proxyLogLevel == 0` and connection without TLS.
```
Comment (lhotari) [inline pulsar-proxy/src/test/resources/proxy.conf:64]: Please reword the explanation to make it easier to read. 

```suggestion
# Enables zero-copy transport of data across network interfaces using the splice system call.
# Zero copy mode cannot be used when TLS is enabled or when proxyLogLevel is > 0
```

File: site2/docs/reference-configuration.md [hunk 5]
```diff
@@ -847,6 +847,7 @@ The [Pulsar proxy](concepts-architecture-overview.md#pulsar-proxy) can be config
 |tokenAudienceClaim| The token audience "claim" name, e.g. "aud". It is used to get the audience from token. If it is not set, the audience is not verified. ||
 | tokenAudience | The token audience stands for this broker. The field `tokenAudienceClaim` of a valid token need contains this parameter.| |
 |haProxyProtocolEnabled | Enable or disable the [HAProxy](http://www.haproxy.org/) protocol. |false|
+|proxyZeroCopyModeEnabled | Enable or disable the `zero-copy` of NIC to NIC. If enabled, we will use [splice](https://man7.org/linux/man-pages/man2/splice.2.html) system call instead of read/write to transport data of NIC to NIC, it only work when `proxyLogLevel == 0` and connection without TLS. | true |
```
Comment (lhotari) [inline site2/docs/reference-configuration.md:850]: Please reword the explanation to make it easier to read. 
```suggestion
|proxyZeroCopyModeEnabled |Enables zero-copy transport of data across network interfaces using the [splice](https://man7.org/linux/man-pages/man2/splice.2.html) system call. Zero copy mode cannot be used when TLS is enabled or when `proxyLogLevel is > 0`. | true |
```

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java [hunk 6]
```diff
@@ -237,6 +237,10 @@ public class ProxyConfiguration implements PulsarConfiguration {
             doc = "Enable or disable the proxy protocol.")
     private boolean haProxyProtocolEnabled;
 
+    @FieldContext(category = CATEGORY_SERVER,
+            doc = "Enable or disable the zero-copy of NIC to NIC.")
```
Comment (lhotari) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java:241]: Improve the doc
```suggestion
            doc = "Enables zero-copy transport of data across network interfaces using the spice. Zero copy mode cannot be used when TLS is enabled or when proxyLogLevel is > 0.")
```

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java [hunk 7]
```diff
@@ -259,6 +271,33 @@ public void channelRead(final ChannelHandlerContext ctx, Object msg) throws Exce
...
+            if (!future.isSuccess()) {
+                future.channel().pipeline().fireExceptionCaught(future.cause());
+            } else {
+                ProxyService.OPS_COUNTER.inc();
+                directProxyHandler.getInboundChannelRequestsRate().recordEvent(SPLICE_BYTES);
+                ProxyService.BYTES_COUNTER.inc(SPLICE_BYTES);
+            }
+        });
+
+        inboundChannel.config().setAutoRead(true);
```
Comment (lhotari) [inline pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java:294]: Doesn't this have to happen in the promise listener after the `spliceTo` operation has completed?

**Review by lhotari — CHANGES_REQUESTED** [review]

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java (hunk 7 above)
Comment (coderzc) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java]: Thanks, I remove the settings of `autoRead`.

**Comment by coderzc** [comment]
@lhotari @eolivelli @merlimat PTAL

**Comment by coderzc** [comment]
@eolivelli @merlimat Please take a look

File: pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lhotari) [reply pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java]: @eolivelli This particular code is only executed once when the handshake completes so this is not a hotspot.

**Review by eolivelli — APPROVED** [review]
Lgtm

File: pulsar-proxy/src/test/resources/proxy.conf [hunk 8]
```diff
@@ -60,6 +60,10 @@ advertisedAddress=
 # Enable or disable the HAProxy protocol.
 haProxyProtocolEnabled=false
 
+# Enables zero-copy transport of data across network interfaces using the splice system call.
+# Zero copy mode cannot be used when TLS is enabled or when proxyLogLevel is > 0.
+proxyZeroCopyModeEnabled=true
```
[the remark this answers is not in the document]
Comment (coderzc) [reply pulsar-proxy/src/test/resources/proxy.conf:65]: ok, thanks. I add test to cover `proxyZeroCopyModeEnabled=false` case

**Comment by coderzc** [comment]
@codelipenghui Can you help me to re-run the failure of the test?

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**